### PR TITLE
[Win] Support backend loading mechanism on Windows

### PR DESCRIPTION
**Pull request by fodinabor**
As some changes to CMake were required as well, this includes and thus supersedes #452.

The main addition on top of #452 is, the dynamic library loading mechanism is also supported on Windows now. For this, the loading / closing functions have been moved to separate functions with integrated error handling, which always return `nullptr`s if something went wrong to simplify the calling functions.

Additionally, the library loading mechanism is relaxed: a new function, which builds a list of unique search paths, is added. One of those paths is the _old_ mechanism, of using the `CMAKE_INSTALL_PREFIX/...` directory. Alternatively, `dladdr` and `GetModuleHandle` are used to determine the position of the `(lib)hipSYCL-rt.{so|dll}` which then is used as base path.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -99,9 +100,19 @@ if(BUILD_CLANG_PLUGIN)
...
-    message(SEND_ERROR "Could not find clang executable")
+    find_program(CLANG_EXECUTABLE_PATH NAMES clang++-${LLVM_VERSION_MAJOR} clang++-${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR} clang++ CACHE STRING)
+    if(CLANG_EXECUTABLE_PATH MATCHES "-NOTFOUND")
+      message(SEND_ERROR "Could not find clang executable")
+    endif()
+  endif()
+  if(WIN32)
+    # plugins link agains clang.exe and not clang++.exe, thus using clang++ would result
+    # in symbol clashes and therefore the plugin not loading.
+    # on Windows it seems, to be fine, as long as clang is able to deduce the language
```
Comment (illuhad) [inline CMakeLists.txt]:
```suggestion
    # on Windows it seems to be fine, as long as clang is able to deduce the language
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -216,25 +260,21 @@ set(SYCLCC_CONFIG_FILE "{
...
-configure_file(
-  ${PROJECT_SOURCE_DIR}/include/hipSYCL/common/config.hpp.in
-  ${PROJECT_BINARY_DIR}/include/hipSYCL/common/config.hpp)
+set(HIPSYCL_INSTALL_CMAKE_DIR
+  "lib/cmake/${PROJECT_NAME}" CACHE PATH "Install path for CMake config files")
 
 install(DIRECTORY include/CL DESTINATION include/ FILES_MATCHING PATTERN "*.hpp")
 install(DIRECTORY include/SYCL DESTINATION include/ FILES_MATCHING PATTERN "*.hpp")
-install(DIRECTORY include/hipSYCL DESTINATION include/ FILES_MATCHING PATTERN "*.hpp")
+install(DIRECTORY include/hipSYCL DESTINATION include/ FILES_MATCHING PATTERN "*.hpp" PATTERN include/hipSYCL/compiler EXCLUDE)
```
Comment (illuhad) [inline CMakeLists.txt]: Why is it important to not install the compiler headers? They are not used when compiling with `syclcc`, but maybe at some point someone would like to build something on top of the hipSYCL infrastructure?

File: bin/syclcc-clang [hunk 3]
```diff
@@ -731,9 +747,13 @@ class hip_invocation:
...
         "-x", "hip",
         "-D__HIPSYCL_ENABLE_HIP_TARGET__",
-        "-D__HIPSYCL_CLANG__",
-        "-fplugin=" + os.path.join(self._hipsycl_lib_path, "libhipSYCL_clang.so")
+        "-D__HIPSYCL_CLANG__"
       ]
+    if sys.platform.startswith('win32'):
+      flags += ["-fplugin=" + os.path.join(self._hipsycl_bin_path, "hipSYCL_clang.dll")]
+    else:
+      flags += ["-fplugin=" + os.path.join(self._hipsycl_lib_path, "libhipSYCL_clang.so")]
```
Comment (illuhad) [inline bin/syclcc-clang]: Maybe pull the plugin path into a property of the `config` class to avoid code duplication?

File: cmake/hipsycl-config.cmake.in [hunk 4]
```diff
@@ -140,7 +140,11 @@ endif()
 # arguments from HIPSYCL_SYCLCC_EXTRA_ARGS and replace GCC or Clang with syclcc in the command line.
 # This is done to keep COMPILE_FLAGS free from Clang-incompatible command line arguments, allowing it to be reused
 # by clang(d)-based tooling and IDEs.
+if(WIN32)
+set(HIPSYCL_SYCLCC_LAUNCH_RULE "python ${HIPSYCL_SYCLCC_LAUNCHER} --launcher-cxx-compiler=${CMAKE_CXX_COMPILER} --launcher-syclcc=\"python*${HIPSYCL_SYCLCC}\" ${HIPSYCL_SYCLCC_EXTRA_ARGS}")
```
Comment (illuhad) [inline cmake/hipsycl-config.cmake.in]: Indentation?

File: cmake/syclcc-launcher [hunk 5]
```diff
@@ -45,7 +45,7 @@ if __name__ == '__main__':
     if arg.startswith(CXX_COMPILER_ARG):
       cxx_compiler_exe = arg[len(CXX_COMPILER_ARG):]
     elif arg.startswith(SYCLCC_ARG):
-      syclcc_exe = arg[len(SYCLCC_ARG):]
+      syclcc_exe = arg[len(SYCLCC_ARG):].split("*")
```
Comment (illuhad) [inline cmake/syclcc-launcher:51]: Can you please elaborate what this does?

File: cmake/syclcc-launcher [hunk 6]
```diff
@@ -66,7 +66,8 @@ if __name__ == '__main__':
   # When invoked with a command line for expected compiler (e.g. clang++), replace with a syclcc invocation.
   # Otherwise, e.g. for invocations of `ar` for linking static libraries, just continue with the command as-is.
   if os.path.samefile(cxx_compiler_exe, command_in):
-    command_line = [syclcc_exe, *syclcc_specific_args, *command_in_args]
+    command_line = syclcc_exe
+    command_line += ([*syclcc_specific_args, *command_in_args])
```
Comment (illuhad) [inline cmake/syclcc-launcher]: Why is this change required?

File: src/compiler/CMakeLists.txt [hunk 7]
```diff
@@ -35,10 +35,16 @@ target_link_libraries(hipSYCL_clang
   ${Boost_LIBRARIES})
 if(APPLE)
     target_link_libraries(hipSYCL_clang -Wl,-undefined -Wl,dynamic_lookup)
+elseif(WIN32)
+	set(CLANG_LIBRARY_DIR ${LLVM_PREFIX_DIR}/lib/)
+	target_link_directories(hipSYCL_clang PRIVATE ${CLANG_LIBRARY_DIR})
+  target_link_libraries(hipSYCL_clang clang)
+  set_target_properties(hipSYCL_clang PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS On)
```
Comment (illuhad) [inline src/compiler/CMakeLists.txt:42]: Indentation seems a bit messed up?

File: src/runtime/CMakeLists.txt [hunk 8]
```diff
@@ -34,8 +34,29 @@ target_include_directories(hipSYCL-rt
...
+  target_link_libraries(hipSYCL-rt PRIVATE dl)
+
+  # ensure actual output and expected are equivalent
+  set_target_properties(hipSYCL-rt PROPERTIES LIBRARY_OUTPUT_NAME ${HIPSYCL_RT_LIBRARY_OUTPUT_NAME})
+  set(HIPSYCL_RT_LIBRARY_OUTPUT_NAME "lib${HIPSYCL_RT_LIBRARY_OUTPUT_NAME}.so")
+endif()
+
+configure_file(
+  ${PROJECT_SOURCE_DIR}/include/hipSYCL/common/config.hpp.in
+  ${PROJECT_BINARY_DIR}/include/hipSYCL/common/config.hpp)
```
Comment (illuhad) [inline src/runtime/CMakeLists.txt]: This file is not only required by the runtime. Why does it need to be configured by the runtime CMakeLists.txt? Can we not also configure in the main CMakeLists?

File: src/runtime/backend_loader.cpp [hunk 9]
```diff
@@ -32,68 +32,131 @@
...
-      HIPSYCL_DEBUG_ERROR << "backend_loader: Could not retrieve backend name"
-                          << err << std::endl;
+      HIPSYCL_DEBUG_ERROR << err << std::endl;
+    }
+  }
+#else
+  if(HMODULE handle = LoadLibraryA(filename.c_str())) {
+    return static_cast<void*>(handle);
+  } else {
+    // too lazy to use FormatMessage bs right now, so look up the error at https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes
```
Comment (illuhad) [inline src/runtime/backend_loader.cpp]: line break ;)

File: src/runtime/backend_loader.cpp [hunk 10]
```diff
@@ -32,68 +32,131 @@
...
+}
 
-  return create_backend_func();
+std::vector<std::filesystem::path> get_plugin_search_paths()
+{
+  std::vector<std::filesystem::path> paths;
+#ifndef _WIN32
+  Dl_info info;
+  if (dladdr(reinterpret_cast<void*>(&get_plugin_search_paths), &info)) {
+      paths.emplace_back(std::filesystem::path{info.dli_fname}.parent_path() / "hipSYCL");
```
Comment (illuhad) [inline src/runtime/backend_loader.cpp]: use 2 spaces as indentation

File: src/runtime/backend_loader.cpp [hunk 11]
```diff
@@ -32,68 +32,131 @@
...
+#else
+  if(HMODULE handle = GetModuleHandleA(HIPSYCL_RT_LIBRARY_NAME))
+  {
+    std::vector<char> pathB(MAX_PATH);
+    if(GetModuleFileNameA(handle, pathB.data(), pathB.size()))
+    {
+      paths.emplace_back(std::filesystem::path{pathB.data()}.parent_path() / "hipSYCL");
+    }
+  }
+  const auto installPrefixedPath = std::filesystem::path{HIPSYCL_INSTALL_PREFIX} / "bin" / "hipSYCL";
```
Comment (illuhad) [inline src/runtime/backend_loader.cpp]: `snake_case`

File: src/runtime/backend_loader.cpp [hunk 12]
```diff
@@ -32,68 +32,131 @@
...
-  return create_backend_func();
+std::vector<std::filesystem::path> get_plugin_search_paths()
+{
+  std::vector<std::filesystem::path> paths;
+#ifndef _WIN32
+  Dl_info info;
+  if (dladdr(reinterpret_cast<void*>(&get_plugin_search_paths), &info)) {
+      paths.emplace_back(std::filesystem::path{info.dli_fname}.parent_path() / "hipSYCL");
+  }
+  const auto installPrefixedPath = std::filesystem::path{HIPSYCL_INSTALL_PREFIX} / "lib" / "hipSYCL";
```
Comment (illuhad) [inline src/runtime/backend_loader.cpp]: `snake_case`

**Review by illuhad — COMMENTED** [review]
Looks good, some remarks :)

File: cmake/syclcc-launcher (hunk 5 above)
Comment (fodinabor) [reply cmake/syclcc-launcher:51]: On windows you can't just call `syclcc`, but instead you need `python syclcc`. To accommodate for this, I added `python*${HIPSYCL_SYCLCC}` in `hipsycl-config.cmake.in`.
The asterisc is required, as we need this to be two different array elements for the subprocess launch, as otherwise `python syclcc` as a filename, would be searched for, which obviously does not exist, so we split at the `*`.
We can't just split at the whitespace, as `${HIPSYCL_SYCLCC}` might be a full path, which on Windows might contain spaces not too uncommonly .. so I chose a character that is not allowed in paths.

File: cmake/syclcc-launcher (hunk 6 above)
Comment (fodinabor) [reply cmake/syclcc-launcher]: `syclcc_exe` now is an array, thus we'd have `[[python, syclcc], arg, arg2]` else.. but we could probably also use the `*` here.. :P

File: src/runtime/CMakeLists.txt (hunk 8 above)
Comment (fodinabor) [reply src/runtime/CMakeLists.txt]: We need the `HIPSYCL_RT_LIBRARY_OUTPUT_NAME` for the config as well, which was not available in the root.
The variable is currently required on Windows for the `GetModuleHandle`, as I did not want to hardcode it in there and to make sure it actually fits the output name, I added this mechanism here.

File: CMakeLists.txt (hunk 2 above)
Comment (fodinabor) [reply CMakeLists.txt]: it's probably not.. this change is actually some months old, so I don't remember, why I thought it was that important..

File: src/runtime/CMakeLists.txt (hunk 8 above)
Comment (fodinabor) [reply src/runtime/CMakeLists.txt]: Tried mitigating this in 257f63c

**Comment by fodinabor** [comment]
@hipSYCL-ci-bot test this on the new infra, please :)

File: cmake/syclcc-launcher (hunk 5 above)
Comment (illuhad) [reply cmake/syclcc-launcher:51]: Ah, I see. I guess we can leave it like that, but maybe add a comment?

Comment (fodinabor) [reply cmake/syclcc-launcher:51]: sure :)

**Review by illuhad — APPROVED** [review]
